development/extensions/tutorial_php_events_and_extending_extensions.rst

```diff
@@ -0,0 +1,325 @@
+loaded, which is especially useful in cases where service priority and/or dependency
+requires they be loaded in a specified order.
+
+Ordered service collections are based on a normal service collection, but the
+collection is sorted with `ksort <http://php.net/ksort>`_. The usage of the
+sorted service collection is nearly the same as the normal service collection,
+except instead of using ``service_collection`` you should use ``ordered_service_collection``:
+
+.. code-block:: yaml
+
+  acme.demo.foobar_collection:
+      class: phpbb\di\ordered_service_collection
+      arguments:
+          - '@service_container'
+      tags:
+          - { name: service_collection, tag: acme.demo.foobar_service }
+
+And adding a service to the ordered service collection:
+
+.. code-block:: yaml
+
+  acme.demo.foobar_foo:
+       class: acme\demo\foobar_foo
+       tags:
+           - { name: acme.demo.foobar_service, order: 1 }
+
+   acme.demo.foobar_bar:
+       class: acme\demo\foobar_bar
+       tags:
+           - { name: acme.demo.foobar_service, order: 2 }
+
+Using the phpBB finder tool
+===========================
+This is probably the least used method because it requires a rigid file and
+directory naming structure, but in doing so it provides the most reliable
+organization of files, so you always can be sure where to look if you want to
+find a certain feature. The extension finder object is used to traverse the
+directory tree to look for files that are located in specific folders and adhere
+to a set of requirements. It is used, for example, to locate migration files,
+both in the core, and in extensions, without those files having to all be
+registered as services.
+
+The ``\phpbb\extension\finder`` is available from the service container as
+``ext.finder`` and can be used as follows. The following example is part of what
+is used to find all routing files, both for extensions and core routes.
+
+.. code-block:: php
+
+  $finder
+      ->directory('config')
+      ->suffix('routing.yml')
+      ->find();
+
+As you can see, you are able to chain method calls together (ending with ``find()``).
+Check out the class definition for more information about the different methods
+that are available (such as specifying a different directory for extensions than
+core files, and getting class names based on files it finds). The return of the
+``find()`` method is an array of file paths that match the given criteria.
+
+For example, assume you have an extension and you need to get a collection of all
+PNG image files from one of its directories. You could use the extension manager
+to load the finder, and traverse the extension's image directory as follows:
+
+.. code-block:: php
+
+  $finder = $extension_manager->get_finder();
+
+  $images = $finder
+      ->extension_suffix('.png')
+      ->extension_directory('/images')
+      ->find_from_extension('demo', $phpbb_root_path . 'ext/acme/demo/');
```

Comment:
`find_from_extension` will search for the files in a specific extension. In 99% of the use cases i think we will prefer `find()` which will search in all extensions.

In 99% of the use case, `suffix()` and `directory()` ill be enough too (they are used when we need to specify different directory or suffix for the core and the extensions. By the way,  if we want to search in the core we need to call core_path() too)